app/controllers/device_controller.rb

```diff
@@ -1,5 +1,59 @@
 class DeviceController < ApplicationController
+  before_action :authorize_request, :only=>[:add_tag, :delete_tag]
+
   def show
     @devices = Device.all
   end
+
+  def raw_data
```

Comment:
Could you move this API back to ReportController or another independent controller? Eventually there will be various rankings, which will be distinguished by the parameters passed to this method. For the front end, each record is still a report, and the front end does not care whether the report is based on device or instance.